### PR TITLE
Kotlin Conversions

### PR DESCRIPTION
**Pull request by hiqua**

### CODE REVIEW

File: uhabits-android/src/main/java/org/isoron/uhabits/widgets/views/EmptyWidgetView.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ */
+package org.isoron.uhabits.widgets.views
+
+import android.content.Context
+import android.view.View
+import android.widget.TextView
+import org.isoron.uhabits.R
+
+class EmptyWidgetView(context: Context?) : HabitWidgetView(context) {
+    private var title: TextView? = null
```
Comment (iSoron) [inline uhabits-android/src/main/java/org/isoron/uhabits/widgets/views/EmptyWidgetView.kt:27]: Please make this `lateinit var` instead

File: uhabits-android/src/main/java/org/isoron/uhabits/widgets/views/GraphWidgetView.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+package org.isoron.uhabits.widgets.views
+
+import android.content.Context
+import android.view.View
+import android.view.ViewGroup
+import android.widget.TextView
+import org.isoron.uhabits.R
+
+class GraphWidgetView(context: Context?, val dataView: View) : HabitWidgetView(context) {
+    private var title: TextView? = null
```
Comment (iSoron) [inline uhabits-android/src/main/java/org/isoron/uhabits/widgets/views/GraphWidgetView.kt:28]: Please make this `lateinit var` instead

**Review by iSoron — CHANGES_REQUESTED** [review]
Thank you for the pull request, @hiqua. This is great!

I have reviewed about half of the files; I will review the remaining ones soon.

**Comment by hiqua** [comment]
Ready for a second review. I'll merge the fixing commits with the corresponding others once we're through.

File: uhabits-android/src/androidTest/java/org/isoron/uhabits/acceptance/AboutTest.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+@LargeTest
+class AboutTest : BaseUserInterfaceTest() {
+    @Test
+    fun shouldDisplayAboutScreen() {
+        launchApp()
+        ListHabitsSteps.clickMenu(ListHabitsSteps.MenuItem.ABOUT)
```
Comment (iSoron) [inline uhabits-android/src/androidTest/java/org/isoron/uhabits/acceptance/AboutTest.kt:37]: Could you static import this method too?

File: uhabits-android/src/androidTest/java/org/isoron/uhabits/acceptance/HabitsTest.kt [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        verifyShowsScreen(CommonSteps.Screen.SHOW_HABIT)
+        clickMenu(ListHabitsSteps.MenuItem.EDIT)
+        verifyShowsScreen(CommonSteps.Screen.EDIT_HABIT)
+        typeName("Take a walk")
+        typeQuestion("Did you take a walk today?")
+        pickColor(10)
+        clickSave()
+        verifyShowsScreen(CommonSteps.Screen.SHOW_HABIT)
+        verifyDisplaysText("Take a walk")
+        CommonSteps.pressBack()
```
Comment (iSoron) [inline uhabits-android/src/androidTest/java/org/isoron/uhabits/acceptance/HabitsTest.kt:143]: Could you static-import this method?

File: uhabits-android/src/androidTest/java/org/isoron/uhabits/acceptance/LinksTest.kt [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+@LargeTest
+class LinksTest : BaseUserInterfaceTest() {
+    @Test
+    @Throws(Exception::class)
+    fun shouldLinkToSourceCode() {
+        launchApp()
+        ListHabitsSteps.clickMenu(ListHabitsSteps.MenuItem.ABOUT)
```
Comment (iSoron) [inline uhabits-android/src/androidTest/java/org/isoron/uhabits/acceptance/LinksTest.kt:38]: Could you static-import this method?

File: uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/ListHabitsRootView.kt [hunk 6]
```diff
@@ -126,17 +124,17 @@ class ListHabitsRootView @Inject constructor(
     override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
         val count = getCheckmarkCount()
         header.buttonCount = count
-        header.setMaxDataOffset(max(MAX_CHECKMARK_COUNT - count, 0))
+        header.setMaxDataOffset(kotlin.math.max(MAX_CHECKMARK_COUNT - count, 0))
```
Comment (iSoron) [inline uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/ListHabitsRootView.kt:127]: Could you static-import `max` and `min`?

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/BaseUnitTest.kt [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+import java.sql.DriverManager
+import java.sql.SQLException
+
+@RunWith(MockitoJUnitRunner::class)
+open class BaseUnitTest {
+    protected open var habitList: HabitList? = null
+    protected var fixtures: HabitFixtures? = null
+    protected var modelFactory: ModelFactory? = null
+    protected var taskRunner: SingleThreadTaskRunner? = null
+    protected open var commandRunner: CommandRunner? = null
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/BaseUnitTest.kt:57]: I think the majority of these are initialized in `setUp`, so they could be `lateinit var` instead

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/BaseUnitTest.kt [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    @Throws(Exception::class)
+    open fun tearDown() {
+        Mockito.validateMockitoUsage()
+        setFixedLocalTime(null)
+        setStartDayOffset(0, 0)
+    }
+
+    fun unixTime(year: Int, month: Int, day: Int): Long {
+        val cal = getStartOfTodayCalendar()
+        cal[year, month, day, 0, 0] = 0
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/BaseUnitTest.kt:98]: It seems like the automatic Kotlin conversion transforms `cal.set(year, month, day, hour, min, sec)` into `cal[year, month, day, hour, min] = sec`. Could you correct this?

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/BaseUnitTest.kt [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+        return cal.timeInMillis
+    }
+
+    open fun unixTime(year: Int, month: Int, day: Int, hour: Int, minute: Int): Long {
+        val cal = getStartOfTodayCalendar()
+        cal[year, month, day, hour] = minute
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/BaseUnitTest.kt:104]: Same here

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/BaseUnitTest.kt [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    @Throws(IOException::class)
+    protected fun openAsset(assetPath: String): InputStream {
+        var `in` = javaClass.getResourceAsStream(assetPath)
+        if (`in` != null) return `in`
+        val pwd = Paths.get(".").toAbsolutePath().normalize().toString()
+        val fullPath = "$pwd/assets/test/$assetPath"
+        val file = File(fullPath)
+        if (file.exists() && file.canRead()) `in` = FileInputStream(file)
+        if (`in` != null) return `in`
+        throw IllegalStateException("asset not found: $fullPath")
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/BaseUnitTest.kt:130]: Could you rename `in` so we don't need backticks?

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/ChangeHabitColorCommandTest.kt [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+    @Test
+    fun testExecute() {
+        checkOriginalColors()
+        command!!.run()
+        checkNewColors()
+    }
+
+    private fun checkNewColors() {
+        for ((color) in selected!!) MatcherAssert.assertThat(
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/ChangeHabitColorCommandTest.kt:55]: Please replace `(color) in selected!!` by `habit in selected!!`, and then `color` by `habit.color`.

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/ChangeHabitColorCommandTest.kt [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+            color,
+            CoreMatchers.equalTo(PaletteColor(0))
+        )
+    }
+
+    private fun checkOriginalColors() {
+        var k = 0
+        for ((color) in selected!!) MatcherAssert.assertThat(
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/ChangeHabitColorCommandTest.kt:63]: Same here

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/CreateHabitCommandTest.kt [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+        model.reminder = Reminder(8, 30, WeekdayList.EVERY_DAY)
+        command = CreateHabitCommand(modelFactory!!, habitList!!, model)
+    }
+
+    @Test
+    fun testExecute() {
+        Assert.assertTrue(habitList!!.isEmpty)
+        command!!.run()
+        MatcherAssert.assertThat(habitList!!.size(), CoreMatchers.equalTo(1))
+        val (_, _, _, _, _, name) = habitList!!.getByPosition(0)
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/CreateHabitCommandTest.kt:49]: Please replace this by `val habit = habitList!!.getByPosition(0)`

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/EditHabitCommandTest.kt [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+import org.isoron.uhabits.core.models.Timestamp
+import org.isoron.uhabits.core.utils.DateUtils.Companion.getTodayWithOffset
+import org.junit.Before
+import org.junit.Test
+
+class EditHabitCommandTest : BaseUnitTest() {
+    private var command: EditHabitCommand? = null
+    private var habit: Habit? = null
+    private var modified: Habit? = null
+    private var today: Timestamp? = null
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/EditHabitCommandTest.kt:35]: Please replace these by `lateinit var`

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/EditHabitCommandTest.kt [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+        modified!!.name = "modified"
+        habitList!!.add(modified!!)
+        today = getTodayWithOffset()
+    }
+
+    @Test
+    fun testExecute() {
+        command = EditHabitCommand(habitList!!, habit!!.id!!, modified!!)
+        val originalScore = habit!!.scores[today!!].value
+        MatcherAssert.assertThat(habit!!.name, Matchers.equalTo("original"))
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/EditHabitCommandTest.kt:56]: Could you static-import `assertThat` and `equalTo`?

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/UnarchiveHabitsCommandTest.kt [hunk 16]
```diff
@@ -16,43 +16,31 @@
...
     @Test
-    public void testExecute()
-    {
-        assertFalse(habit.isArchived());
-        command.run();
-        assertTrue(habit.isArchived());
+    fun testExecuteUndoRedo() {
+        Assert.assertTrue(habit.isArchived)
+        command!!.run()
+        Assert.assertFalse(habit.isArchived)
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/UnarchiveHabitsCommandTest.kt:44]: Please static-import `assert` methods

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/database/migrations/Version22Test.kt [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+import org.isoron.uhabits.core.test.HabitFixtures
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.ExpectedException
+
+class Version22Test : BaseUnitTest() {
+    @get:Rule
+    var exception = ExpectedException.none()!!
+    private var db: Database? = null
+    private var helper: MigrationHelper? = null
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/database/migrations/Version22Test.kt:37]: Please replace by `lateinit var`

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/io/HabitsCSVExporterTest.kt [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+class HabitsCSVExporterTest : BaseUnitTest() {
+    private var baseDir: File? = null
+    @Before
+    @Throws(Exception::class)
+    override fun setUp() {
+        super.setUp()
+        habitList!!.add(fixtures!!.createShortHabit())
+        habitList!!.add(fixtures!!.createEmptyHabit())
+        baseDir = Files.createTempDirectory("csv").toFile()
+        Assert.assertNotNull(baseDir)
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/io/HabitsCSVExporterTest.kt:44]: Please static-import assert methods

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/io/HabitsCSVExporterTest.kt [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+    //                        "2015-01-19,0.0706\n" +
+    //                        "2015-01-18,0.0515\n" +
+    //                        "2015-01-17,0.0315\n" +
+    //                        "2015-01-16,0.0107\n";
+    //
+    //        StringWriter writer = new StringWriter();
+    //        habit.getScores().writeCSV(writer);
+    //
+    //        assertThat(writer.toString(), equalTo(expectedCSV));
+    //    }
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/io/HabitsCSVExporterTest.kt:118]: Please delete this. Also, feel free to delete any other commented-out tests.

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/io/ImportTest.kt [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+    @Throws(IOException::class)
+    fun testLoopDB() {
+        importFromFile("loop.db")
+        MatcherAssert.assertThat(habitList!!.size(), IsEqual.equalTo(9))
+        val habit = habitList!!.getByPosition(0)
+        MatcherAssert.assertThat(habit.name, IsEqual.equalTo("Wake up early"))
+        MatcherAssert.assertThat(habit.frequency, IsEqual.equalTo(Frequency.THREE_TIMES_PER_WEEK))
+        Assert.assertTrue(isChecked(habit, 2016, 3, 14))
+        Assert.assertTrue(isChecked(habit, 2016, 3, 16))
+        TestCase.assertFalse(isChecked(habit, 2016, 3, 17))
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/io/ImportTest.kt:69]: Please static-import assert methods

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/io/ImportTest.kt [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+        MatcherAssert.assertThat(h.name, IsEqual.equalTo("Vegan"))
+        Assert.assertTrue(isChecked(h, 2016, 1, 24))
+        Assert.assertTrue(isChecked(h, 2016, 2, 5))
+        Assert.assertTrue(isChecked(h, 2016, 3, 18))
+        TestCase.assertFalse(isChecked(h, 2016, 3, 14))
+    }
+
+    private fun isChecked(h: Habit, year: Int, month: Int, day: Int): Boolean {
+        val date = getStartOfTodayCalendar()
+        date[year, month - 1] = day
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/io/ImportTest.kt:111]: Please replace this by `date.set(year, month-1, day)`

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+import java.io.IOException
+import java.io.StringWriter
+import java.util.ArrayList
+
+class HabitListTest : BaseUnitTest() {
+    @get:Rule
+    var thrown = ExpectedException.none()!!
+    private var habitsArray: ArrayList<Habit>? = null
+    private var activeHabits: HabitList? = null
+    private var reminderHabits: HabitList? = null
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt:38]: Please replace by `lateinit var`

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt [hunk 23]
```diff
@@ -0,0 +1,307 @@
...
+                .setReminderRequired(true)
+                .build()
+        )
+    }
+
+    @Test
+    fun testSize() {
+        MatcherAssert.assertThat(habitList!!.size(), CoreMatchers.equalTo(10))
+        MatcherAssert.assertThat(activeHabits!!.size(), CoreMatchers.equalTo(6))
+        MatcherAssert.assertThat(reminderHabits!!.size(), CoreMatchers.equalTo(4))
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt:66]: Please use static imports in `assertThat` and `equalTo`. Same for the other usages in this file.

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt [hunk 24]
```diff
@@ -0,0 +1,307 @@
...
+    fun testGetById_withInvalidId() {
+        Assert.assertNull(habitList!!.getById(100L))
+    }
+
+    @Test
+    fun testOrdering() {
+        val h1 = fixtures!!.createEmptyHabit()
+        h1.name = "A Habit"
+        h1.color = PaletteColor(2)
+        h1.position = 1
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt:118]: Could you use `apply` to make this look nicer?
```kotlin
val h1 = fixtures.createEmptyHabit().apply {
    name = "A Habit"
    color = [...]
}
```
An alternative would be making `name`, `color`, `position` etc arguments of `createEmptyHabit`, but provide default values.

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt [hunk 25]
```diff
@@ -0,0 +1,307 @@
...
+        h3.position = 0
+        val h4 = fixtures!!.createEmptyHabit()
+        h4.name = "D Habit"
+        h4.color = PaletteColor(1)
+        h4.position = 2
+        val list = modelFactory!!.buildHabitList()
+        list.add(h3)
+        list.add(h1)
+        list.add(h4)
+        list.add(h2)
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt:135]: `apply` could also be used here

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt [hunk 26]
```diff
@@ -0,0 +1,307 @@
...
+            intArrayOf(0, 1, 5, 2, 4, 6, 7, 3, 8, 9),
+            intArrayOf(0, 1, 5, 2, 4, 6, 7, 8, 3, 9)
+        )
+        for (i in operations.indices) {
+            val fromHabit = habitsArray!![operations[i][0]]
+            val toHabit = habitsArray!![operations[i][1]]
+            habitList!!.reorder(fromHabit, toHabit)
+            val actualSequence = IntArray(10)
+            for (j in 0..9) {
+                val (_, _, _, id, _, _, position) = habitList!!.getByPosition(j)
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt:190]: Please change to `val habit = habitList!!.getByPosition(j)`

**Review by iSoron — CHANGES_REQUESTED** [review]

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt [hunk 27]
```diff
@@ -0,0 +1,301 @@
...
+    }
+
+    @Test
+    fun testGetByPosition() {
+        assertThat(
+            habitList.getByPosition(0),
+            CoreMatchers.equalTo(
+                habitsArray[0]
+            )
+        )
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt:79]: Please put these in a single line. I think ktlint breaks lines, but doesn't merge them back.

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/ui/screens/habits/list/ListHabitsMenuBehaviorTest.kt [hunk 28]
```diff
@@ -0,0 +1,214 @@
...
+        whenever(adapter.primaryOrder).thenReturn(HabitList.Order.BY_NAME_ASC)
+        behavior.onSortByStatus()
+        verify(adapter).primaryOrder = orderCaptor.capture()
+        verify(adapter).setSecondaryOrder(
+            secondaryOrderCaptor.capture()
+        )
+        assertThat(
+            orderCaptor.lastValue,
+            equalTo(HabitList.Order.BY_STATUS_ASC)
+        )
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/ui/screens/habits/list/ListHabitsMenuBehaviorTest.kt:141]: I think these could be in a single line. There are many other similar cases in this file.

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/utils/DateUtilsTest.kt [hunk 29]
```diff
@@ -0,0 +1,479 @@
...
+    @Test
+    fun testTruncate_dayOfWeek() {
+        val field = DateUtils.TruncateField.WEEK_NUMBER
+        var expected = unixTime(2015, Calendar.JANUARY, 11)
+        var t0 = unixTime(2015, Calendar.JANUARY, 11)
+        var t1 = unixTime(2015, Calendar.JANUARY, 16)
+        var t2 = unixTime(2015, Calendar.JANUARY, 17)
+        assertThat(truncate(field, t0, firstWeekday), CoreMatchers.equalTo(expected))
+        assertThat(truncate(field, t1, firstWeekday), CoreMatchers.equalTo(expected))
+        assertThat(truncate(field, t2, firstWeekday), CoreMatchers.equalTo(expected))
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/utils/DateUtilsTest.kt:69]: Please use static imports

**Review by iSoron — CHANGES_REQUESTED** [review]
Just a few minor changes requested. Hopefully we can merge it after these are fixed.

**Comment by hiqua** [comment]
I've removed the assertNonNull and replaced them with `!!`. These are bound to disappear completely anyway, once we've got rid of (most of) the nullable types in kotlin.

**Comment by hiqua** [comment]
@iSoron can you update https://github.com/iSoron/uhabits/issues/685 once we're done with this PR?

**Review by iSoron — APPROVED** [review]

**Comment by iSoron** [comment]
Merged! Thanks for this PR, @hiqua. I'm sure it was a lot of work. I'll update #685 soon.

**Pull request merged**